bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesPageInfoProvider.java

```diff
@@ -23,42 +23,88 @@
 import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.json.JSONArray;
 import org.apache.sling.commons.json.JSONException;
 import org.apache.sling.commons.json.JSONObject;
+import org.apache.sling.jcr.api.SlingRepository;
 import org.apache.sling.jcr.base.util.AccessControlUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.jcr.LoginException;
+import javax.jcr.Node;
 import javax.jcr.RepositoryException;
 import javax.jcr.Session;
+import javax.jcr.observation.Event;
+import javax.jcr.observation.EventIterator;
+import javax.jcr.observation.EventListener;
+import javax.jcr.observation.ObservationManager;
 import javax.jcr.security.AccessControlManager;
 import javax.jcr.security.Privilege;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 /**
  * PageInfoProvider which indicates that shared component properties
- * are enabled.  Note that this provider requires Page Root Provider to
- * be configured.
- *
+ * are enabled, and for which component types.  Note that this provider
+ * requires Page Root Provider to be configured.
+ * <p>
  * https://docs.adobe.com/docs/en/cq/5-6-1/developing/pageinfo.html#Creating a Page Information Provider
  */
 @org.apache.felix.scr.annotations.Component
 @Service
-public class SharedComponentPropertiesPageInfoProvider implements PageInfoProvider {
+public class SharedComponentPropertiesPageInfoProvider implements PageInfoProvider, EventListener {
     private static final Logger log = LoggerFactory.getLogger(SharedComponentPropertiesPageInfoProvider.class);
 
+    private static final String SERVICE_NAME = "shared-component-props";
+
     @Reference
     private PageRootProvider pageRootProvider;
 
     @Reference
     private SharedComponentProperties sharedComponentProperties;
 
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+
+    @Reference
+    private SlingRepository repository;
+
+    private Session respositorySession;
+
+    private ObservationManager observationManager;
+
+    private Map<String, List<Boolean>> componentsWithSharedProperties;
+
+    private Calendar scheduledSharedComponentsMapUpdate = null;
+
+    /**
+     * Add a "sharedComponentProperties" section to pageInfo so that JS libs in the authoring interface
+     * can determine whether or not to enable shared/global properties for a component on a site.
+     */
     @Override
     public void updatePageInfo(SlingHttpServletRequest request, JSONObject info, Resource resource)
             throws JSONException {
+        if (scheduledSharedComponentsMapUpdate != null && Calendar.getInstance().after(scheduledSharedComponentsMapUpdate)) {
+            scheduledSharedComponentsMapUpdate = null;
```

Comment:
Understood that this is not 100% threadsafe and that 2 exactly concurrent requests could trigger this to run twice, but the operation runs in well less than a second so not worried about it.